### DISCUSSION TITLE
Semantic functions

### DISCUSSION THREAD

**Discussion by ConsciousCode** [daveshap/ACE_Framework#23] [General]
I'm excited to see the demos for the ACE framework and where it leads. This sort of high-level systemic model makes it a lot easier to conceptualize and try out my own experiments using it.

I have a feeling my [servitor](https://github.com/ConsciousCode/servitor) library, or at least its concept, might be helpful in building the different layers. It's based on Microsoft's [Semantic Kernels](https://github.com/microsoft/semantic-kernel), but stripped down to what I perceive to be a potential new programming abstraction - "semantic functions" - which is simply a function defined in natural language with discrete inputs and outputs. Thus, rather than treat LLMs as conversational agents which require careful curation of input context and allowable responses, they're used as simple language models with predictable outputs. For instance, using my library's API:

```python
@semantic
def task_select(priorities: str, tasks: set[str]) -> list[str]:
    '''Given the priorities set by upstream cognitive agents and a set of tasks to be accomplished, return a task list sorted by priority.'''

tasklist = task_select(southbound, taskset)
task_prosecution(southbound, tasklist)
# ...
```

(Note that the library includes additional configurable internal prompting and per-annotation switching of hyperparameters and model).

It's meant for interoperability with the normal programming environment, which may be useful for the more technical layers, but is flexible enough to receive and output ordinary natural language. The LLM is given the function signature it's meant to execute, so you could eg create a dummy class `NaturalLanguage` or just leave it unannotated and omit the postprocessing steps, simply returning the string generation. If for some reason this isn't helpful, I'd also like to know why too since that may provide insights which can inform my own experiments.

**Comment by daveshap** [comment]
This is similar to an idea I cooked up much earlier this year, but it was around finetuning models to provide individual "cognitive functions." My original idea was to finetune lightweight models to provide basic cognitive capabilities like brainstorming or choosing individual options. This is a good direction to go in. 

By creating a reusable toolbox, you make it easier to build a "plug and play" architecture. Thanks
